internal/profiling/profiling.go

```diff
@@ -0,0 +1,213 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// The profiling package contains two logical components: buffer.go and
+// profiling.go. The former implements a circular buffer (a.k.a. ring buffer)
+// in a lock-free manner using atomics. This ring buffer is used by
+// profiling.go to store various statistics. For example, StreamStats is a
+// circular buffer of Stat objects, each of which is comprised of Timers.
+//
+// This abstraction is designed to accomodate more stats in the future; for
+// example, if one wants to profile the load balancing layer, which is
+// independent of RPC queries, a separate CircularBuffer can be used.
+//
+// Note that the circular buffer simply takes any interface{}. In the future,
+// more types of measurements (such as the number of memory allocations) could
+// be measured, which might require a different type of object being pushed
+// into the circular buffer.
+
+package profiling
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// 0 or 1 representing profiling off and on, respectively. Use IsEnabled and
+// SetEnabled to get and set this in a safe manner.
+var profilingEnabled uint32
+
+// IsEnabled returns whether or not profiling is enabled.
+func IsEnabled() bool {
+	return atomic.LoadUint32(&profilingEnabled) > 0
+}
+
+// SetEnabled turns profiling on and off.
+//
+// Note that it is impossible to enable profiling for one server and leave it
+// turned off for another. This is intentional and by design -- if the status
+// of profiling was server-specific, clients wouldn't be able to profile
+// themselves. As a result, SetEnabled turns profiling on and off for all
+// servers and clients in the binary. Each stat will be, however, tagged with
+// whether it's a client stat or a server stat; so you should be able to filter
+// for the right type of stats in post-processing.
+// SetEnabled is the internal
+func SetEnabled(enabled bool) {
+	if enabled {
+		atomic.StoreUint32(&profilingEnabled, 1)
+	} else {
+		atomic.StoreUint32(&profilingEnabled, 0)
+	}
+}
+
+// A Timer represents the wall-clock beginning and ending of a logical
+// operation.
+type Timer struct {
```

Comment:
The `Timer` struct has the same fields (functionally) as that of the `TimerProto`. Would it make sense to not define this type at all, and just use the protos here as well?

Reply:
I did that originally, but I faced two disadvantages with using the same struct:

 - `TimerProto` has a few extra fields, which amount to at least 8-10 additional bytes. We want to keep the memory footprint of profiling be as low as possible.

 - `time.Time` does not translate well to protobuf-lingo. There's a time struct that can be imported from Google's protobuf extensions, but internally, it's basically a `sec` field and a `nsec` field, which we end up doing any way. To translate from `time.Now()` (local time) to a UTC sec, nsec pair, we need to do more computation at `Timer` creation time (and `Egress`, of course), which we'd like to avoid.

Reply:
Currently, it takes about ~50 bytes to store a stat. Using `TimerProto` would effectively mean 16% fewer stats for the same amount of memory used.

Reply:
Resolved offline.